File: funtime/lib/breakout-rules.js
/* ============================================================
   breakout-rules.js — the rules of Breakout

   A ball bounces around a box. You slide a paddle along the bottom to keep it
   in play, and every brick it touches disappears. Clear them all to win.

   Unlike the board games, nothing here sits on a grid: the ball has a real
   position in PIXELS and a speed in pixels per second, so it can be anywhere.
   ============================================================ */

const FIELD_WIDTH = 320;
const FIELD_HEIGHT = 420;

const BRICK_COLUMNS = 7;
const BRICK_ROWS = 5;
const BRICK_TOP = 44;
const BRICK_HEIGHT = 18;
const BRICK_GAP = 4;
const BRICK_SIDE_MARGIN = 8;
const BRICK_WIDTH = (FIELD_WIDTH - BRICK_SIDE_MARGIN * 2 - BRICK_GAP * (BRICK_COLUMNS - 1)) / BRICK_COLUMNS;

const PADDLE_WIDTH = 74;
const PADDLE_HEIGHT = 12;
const PADDLE_Y = FIELD_HEIGHT - 30;
const PADDLE_SPEED = 340;      /* pixels per second */

const BALL_RADIUS = 6;
const START_LIVES = 3;

/**
 * brickIndex — turn a column and row into a place in the list.
 * INPUT: column, row. OUTPUT: the position in the flat list of bricks.
 */
function brickIndex(column, row) {
    return row * BRICK_COLUMNS + column;
}

/**
 * brickRect — where one brick sits, as a rectangle.
 *
 * INPUT:  column, row
 * OUTPUT: { x, y, width, height } in pixels
 *
 * ALGORITHM: start at the side margin, skip a brick plus a gap for every
 *            column, and do the same downwards for the rows.
 */
function brickRect(column, row) {
    return {
        x: BRICK_SIDE_MARGIN + column * (BRICK_WIDTH + BRICK_GAP),
        y: BRICK_TOP + row * (BRICK_HEIGHT + BRICK_GAP),
        width: BRICK_WIDTH,
        height: BRICK_HEIGHT
    };
}

/**
 * hitsRect — is the ball touching this rectangle?
 *
 * INPUT:  ball — { x, y } the middle of the ball. rect — { x, y, width, height }.
 * OUTPUT: true if they overlap
 *
 * ALGORITHM (the closest-point trick): find the point of the rectangle that is
 *            nearest the middle of the ball by clamping the ball's x and y to
 *            the rectangle's edges. If that point is closer than the ball's
 *            radius, they are touching.
 */
function hitsRect(ball, rect) {
    const nearestX = Math.max(rect.x, Math.min(ball.x, rect.x + rect.width));
    const nearestY = Math.max(rect.y, Math.min(ball.y, rect.y + rect.height));
    const gapX = ball.x - nearestX;
    const gapY = ball.y - nearestY;
    return gapX * gapX + gapY * gapY <= BALL_RADIUS * BALL_RADIUS;
}

/** paddleRect — the paddle as a rectangle, so the same test works on it. */
function paddleRect(state) {
    return { x: state.paddleX, y: PADDLE_Y, width: PADDLE_WIDTH, height: PADDLE_HEIGHT };
}

/** createBricks — a full wall of bricks, all unbroken. */
function createBricks() {
    const bricks = [];
    for (let i = 0; i < BRICK_COLUMNS * BRICK_ROWS; i++) {
        bricks.push(true);
    }
    return bricks;
}

/** createGame — start a brand-new game. */
function createGame() {
    return {
        bricks: createBricks(),
        paddleX: (FIELD_WIDTH - PADDLE_WIDTH) / 2,
        ball: { x: FIELD_WIDTH / 2, y: PADDLE_Y - 40, dx: 150, dy: -230 },
        steering: 0,
        lives: START_LIVES,
        score: 0,
        level: 1,
        isOver: false,
        isWon: false,
        isPaused: true
    };
}

/**
 * movePaddle — slide the paddle sideways.
 *
 * INPUT:  state — the game. seconds — how long this frame took.
 * OUTPUT: nothing; it changes state.paddleX
 *
 * ALGORITHM: move by steering × PADDLE_SPEED × seconds, then keep the paddle
 *            on the field by clamping it between 0 and the right-hand edge.
 */
function movePaddle(state, seconds) {
    let x = state.paddleX + state.steering * PADDLE_SPEED * seconds;
    x = Math.max(0, Math.min(FIELD_WIDTH - PADDLE_WIDTH, x));
    state.paddleX = x;
}

/**
 * bounceOffWalls — keep the ball inside the box.
 *
 * INPUT:  ball — { x, y, dx, dy }
 * OUTPUT: nothing; it changes the ball
 *
 * ALGORITHM: if the ball has reached the left or right wall, push it back to
 *            the edge and flip dx. Do the same with dy at the ceiling. The
 *            floor is NOT a wall — that is where you lose a life.
 */
function bounceOffWalls(ball) {
    if (ball.x < BALL_RADIUS) {
        ball.x = BALL_RADIUS;
        ball.dx = -ball.dx;
    }
    if (ball.x > FIELD_WIDTH - BALL_RADIUS) {
        ball.x = FIELD_WIDTH - BALL_RADIUS;
        ball.dx = -ball.dx;
    }
    if (ball.y < BALL_RADIUS) {
        ball.y = BALL_RADIUS;
        ball.dy = -ball.dy;
    }
}

/**
 * bounceOffPaddle — bat the ball back into play.
 *
 * INPUT:  state — the game
 * OUTPUT: true if the paddle hit the ball
 *
 * ALGORITHM: only when the ball is falling (dy > 0) and touching the paddle.
 *            Send it back up, and steer it by WHERE it hit: the middle sends
 *            it straight up, the edges send it off at an angle. That is what
 *            gives the player control instead of a game of luck.
 */
function bounceOffPaddle(state) {
    if (state.ball.dy <= 0 || !hitsRect(state.ball, paddleRect(state))) {
        return false;
    }
    const middle = state.paddleX + PADDLE_WIDTH / 2;
    const offset = (state.ball.x - middle) / (PADDLE_WIDTH / 2);
    state.ball.dy = -Math.abs(state.ball.dy);
    state.ball.dx = offset * 240;
    state.ball.y = PADDLE_Y - BALL_RADIUS;
    return true;
}

/**
 * breakBricks — knock out the brick the ball is touching.
 *
 * INPUT:  state — the game
 * OUTPUT: true if a brick was broken
 *
 * ALGORITHM: look at every brick that is still there; the first one the ball
 *            touches is removed, the ball bounces back the way it came, and
 *            the score goes up. Stop after one brick — a ball should not
 *            tunnel through a whole row in a single frame.
 */
function breakBricks(state) {
    for (let row = 0; row < BRICK_ROWS; row++) {
        for (let column = 0; column < BRICK_COLUMNS; column++) {
            const index = brickIndex(column, row);
            if (state.bricks[index] && hitsRect(state.ball, brickRect(column, row))) {
                state.bricks[index] = false;
                state.ball.dy = -state.ball.dy;
                state.score = state.score + scoreForBrick(row);
                if (bricksLeft(state) === 0) {
                    state.isWon = true;
                    state.isOver = true;
                }
                return true;
            }
        }
    }
    return false;
}

/**
 * scoreForBrick — how many points is one brick worth?
 * INPUT: row — 0 is the top row. OUTPUT: points.
 * ALGORITHM: the higher the row, the more it is worth: 50 down to 10.
 */
function scoreForBrick(row) {
    return (BRICK_ROWS - row) * 10;
}

/** bricksLeft — how many bricks are still standing? */
function bricksLeft(state) {
    let count = 0;
    for (let i = 0; i < state.bricks.length; i++) {
        if (state.bricks[i]) { count = count + 1; }
    }
    return count;
}

/** resetBall — put the ball back above the paddle after losing a life. */
function resetBall(state) {
    state.ball = { x: FIELD_WIDTH / 2, y: PADDLE_Y - 40, dx: 150, dy: -230 };
    state.isPaused = true;
}

/**
 * updateGame — one frame of the game.
 *
 * INPUT:  state, elapsedMs
 * OUTPUT: nothing
 *
 * ALGORITHM: move the paddle, move the ball, bounce it off the walls, the
 *            paddle and the bricks, and take a life if it falls off the
 *            bottom. Losing the last life ends the game.
 */
function updateGame(state, elapsedMs) {
    if (state.isOver || state.isPaused) {
        return;
    }
    const seconds = elapsedMs / 1000;

    movePaddle(state, seconds);

    state.ball.x = state.ball.x + state.ball.dx * seconds;
    state.ball.y = state.ball.y + state.ball.dy * seconds;

    bounceOffWalls(state.ball);
    bounceOffPaddle(state);
    breakBricks(state);

    if (state.ball.y > FIELD_HEIGHT + BALL_RADIUS) {
        state.lives = state.lives - 1;
        if (state.lives <= 0) {
            state.lives = 0;
            state.isOver = true;
        } else {
            resetBall(state);
        }
    }
}

/** togglePause — freeze or unfreeze the game. */
function togglePause(state) {
    if (!state.isOver) { state.isPaused = !state.isPaused; }
}

/** actionForKey — turn a keyboard key into an action name, or null. */
function actionForKey(key) {
    const k = String(key).toLowerCase();

    if (k === 'arrowleft' || k === 'a') { return 'left'; }
    if (k === 'arrowright' || k === 'd') { return 'right'; }
    if (k === ' ' || k === 'spacebar' || k === 'p') { return 'pause'; }
    if (k === 'r') { return 'restart'; }

    return null;
}

File: funtime/pylib/breakout_rules.py
"""breakout_rules.py - the rules of Breakout, in Python.

A ball bounces around a box; you slide a paddle along the bottom to keep it in
play and knock out every brick.

Nothing here sits on a grid: the ball has a real position in PIXELS and a
speed in pixels per second.
"""

FIELD_WIDTH = 320
FIELD_HEIGHT = 420

BRICK_COLUMNS = 7
BRICK_ROWS = 5
BRICK_TOP = 44
BRICK_HEIGHT = 18
BRICK_GAP = 4
BRICK_SIDE_MARGIN = 8
BRICK_WIDTH = (FIELD_WIDTH - BRICK_SIDE_MARGIN * 2 - BRICK_GAP * (BRICK_COLUMNS - 1)) / BRICK_COLUMNS

PADDLE_WIDTH = 74
PADDLE_HEIGHT = 12
PADDLE_Y = FIELD_HEIGHT - 30
PADDLE_SPEED = 340

BALL_RADIUS = 6
START_LIVES = 3


def brick_index(column, row):
    """Turn a column and row into a place in the list of bricks."""
    return row * BRICK_COLUMNS + column


def brick_rect(column, row):
    """Where one brick sits, as a rectangle.

    INPUT:  column, row. OUTPUT: a dict with x, y, width, height.
    ALGORITHM: start at the margin, then skip a brick plus a gap for every
    column, and the same downwards for the rows.
    """
    return {
        "x": BRICK_SIDE_MARGIN + column * (BRICK_WIDTH + BRICK_GAP),
        "y": BRICK_TOP + row * (BRICK_HEIGHT + BRICK_GAP),
        "width": BRICK_WIDTH,
        "height": BRICK_HEIGHT,
    }


def hits_rect(ball, rect):
    """Is the ball touching this rectangle?

    INPUT:  ball - a dict with x and y (the middle). rect - x, y, width, height.
    OUTPUT: True if they overlap.
    ALGORITHM (the closest-point trick): squeeze the ball's middle onto the
    rectangle to find the nearest point on it, then see whether that point is
    within one radius.
    """
    nearest_x = max(rect["x"], min(ball["x"], rect["x"] + rect["width"]))
    nearest_y = max(rect["y"], min(ball["y"], rect["y"] + rect["height"]))
    gap_x = ball["x"] - nearest_x
    gap_y = ball["y"] - nearest_y
    return gap_x * gap_x + gap_y * gap_y <= BALL_RADIUS * BALL_RADIUS


def paddle_rect(state):
    """The paddle as a rectangle, so the same test works on it."""
    return {"x": state["paddle_x"], "y": PADDLE_Y, "width": PADDLE_WIDTH, "height": PADDLE_HEIGHT}


def create_bricks():
    """A full wall of bricks, all unbroken."""
    return [True] * (BRICK_COLUMNS * BRICK_ROWS)


def create_game():
    """Start a brand-new game."""
    return {
        "bricks": create_bricks(),
        "paddle_x": (FIELD_WIDTH - PADDLE_WIDTH) / 2,
        "ball": {"x": FIELD_WIDTH / 2, "y": PADDLE_Y - 40, "dx": 150, "dy": -230},
        "steering": 0,
        "lives": START_LIVES,
        "score": 0,
        "level": 1,
        "is_over": False,
        "is_won": False,
        "is_paused": True,
    }


def move_paddle(state, seconds):
    """Slide the paddle sideways, keeping it on the field.

    ALGORITHM: move by steering x PADDLE_SPEED x seconds, then clamp.
    """
    x = state["paddle_x"] + state["steering"] * PADDLE_SPEED * seconds
    state["paddle_x"] = max(0, min(FIELD_WIDTH - PADDLE_WIDTH, x))


def bounce_off_walls(ball):
    """Keep the ball inside the box.

    ALGORITHM: at the left, right or top edge, push the ball back to the edge
    and flip its direction. The floor is NOT a wall - that is where you lose a
    life.
    """
    if ball["x"] < BALL_RADIUS:
        ball["x"] = BALL_RADIUS
        ball["dx"] = -ball["dx"]
    if ball["x"] > FIELD_WIDTH - BALL_RADIUS:
        ball["x"] = FIELD_WIDTH - BALL_RADIUS
        ball["dx"] = -ball["dx"]
    if ball["y"] < BALL_RADIUS:
        ball["y"] = BALL_RADIUS
        ball["dy"] = -ball["dy"]


def bounce_off_paddle(state):
    """Bat the ball back into play.

    OUTPUT: True if the paddle hit the ball.
    ALGORITHM: only when the ball is falling and touching the paddle. Send it
    back up, and steer it by WHERE it hit - the middle sends it straight up,
    the edges send it away at an angle.
    """
    if state["ball"]["dy"] <= 0 or not hits_rect(state["ball"], paddle_rect(state)):
        return False
    middle = state["paddle_x"] + PADDLE_WIDTH / 2
    offset = (state["ball"]["x"] - middle) / (PADDLE_WIDTH / 2)
    state["ball"]["dy"] = -abs(state["ball"]["dy"])
    state["ball"]["dx"] = offset * 240
    state["ball"]["y"] = PADDLE_Y - BALL_RADIUS
    return True


def score_for_brick(row):
    """How many points is one brick worth? The higher the row, the more."""
    return (BRICK_ROWS - row) * 10


def bricks_left(state):
    """How many bricks are still standing?"""
    return sum(1 for brick in state["bricks"] if brick)


def break_bricks(state):
    """Knock out the brick the ball is touching.

    OUTPUT: True if a brick was broken.
    ALGORITHM: the first brick the ball touches is removed, the ball bounces
    back, and the score goes up. Stop after ONE brick, so the ball cannot
    tunnel through a whole row in a single frame.
    """
    for row in range(BRICK_ROWS):
        for column in range(BRICK_COLUMNS):
            index = brick_index(column, row)
            if state["bricks"][index] and hits_rect(state["ball"], brick_rect(column, row)):
                state["bricks"][index] = False
                state["ball"]["dy"] = -state["ball"]["dy"]
                state["score"] += score_for_brick(row)
                if bricks_left(state) == 0:
                    state["is_won"] = True
                    state["is_over"] = True
                return True
    return False


def reset_ball(state):
    """Put the ball back above the paddle after losing a life."""
    state["ball"] = {"x": FIELD_WIDTH / 2, "y": PADDLE_Y - 40, "dx": 150, "dy": -230}
    state["is_paused"] = True


def update_game(state, elapsed_ms):
    """One frame of the game."""
    if state["is_over"] or state["is_paused"]:
        return
    seconds = elapsed_ms / 1000

    move_paddle(state, seconds)

    state["ball"]["x"] += state["ball"]["dx"] * seconds
    state["ball"]["y"] += state["ball"]["dy"] * seconds

    bounce_off_walls(state["ball"])
    bounce_off_paddle(state)
    break_bricks(state)

    if state["ball"]["y"] > FIELD_HEIGHT + BALL_RADIUS:
        state["lives"] -= 1
        if state["lives"] <= 0:
            state["lives"] = 0
            state["is_over"] = True
        else:
            reset_ball(state)


def toggle_pause(state):
    """Freeze or unfreeze the game."""
    if not state["is_over"]:
        state["is_paused"] = not state["is_paused"]


def action_for_key(key):
    """Turn a keyboard key into an action name, or None."""
    keys = {
        "arrowleft": "left", "a": "left",
        "arrowright": "right", "d": "right",
        " ": "pause", "spacebar": "pause", "p": "pause",
        "r": "restart",
    }
    return keys.get(str(key).lower())
